File: Java/Enterprise/DomainModel/Customer.java
package Enterprise.DomainModel;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

/**
 * Domain Model pattern - incorporates both behavior and data.
 * Rich domain object with business logic.
 */
public class Customer {
    private Long id;
    private String name;
    private String email;
    private CustomerType type;
    private List<Order> orders;
    private BigDecimal creditLimit;

    public Customer(String name, String email, CustomerType type) {
        this.name = name;
        this.email = email;
        this.type = type;
        this.orders = new ArrayList<>();
        this.creditLimit = type.getDefaultCreditLimit();
    }

    /**
     * Business logic: Can customer place this order?
     */
    public boolean canPlaceOrder(BigDecimal orderAmount) {
        BigDecimal totalOutstanding = calculateOutstandingBalance();
        BigDecimal projectedBalance = totalOutstanding.add(orderAmount);
        return projectedBalance.compareTo(creditLimit) <= 0;
    }

    /**
     * Business logic: Place an order with validation.
     */
    public Order placeOrder(List<OrderLine> items) {
        BigDecimal orderTotal = items.stream()
                .map(OrderLine::getTotal)
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        if (!canPlaceOrder(orderTotal)) {
            throw new IllegalStateException("Order exceeds credit limit");
        }

        Order order = new Order(this, items);
        orders.add(order);
        return order;
    }

    /**
     * Business logic: Calculate outstanding balance.
     */
    public BigDecimal calculateOutstandingBalance() {
        return orders.stream()
                .filter(order -> !order.isPaid())
                .map(Order::getTotal)
                .reduce(BigDecimal.ZERO, BigDecimal::add);
    }

    /**
     * Business logic: Get discount based on customer type.
     */
    public BigDecimal getDiscountPercentage() {
        return type.getDiscountPercentage();
    }

    /**
     * Business logic: Increase credit limit based on payment history.
     */
    public void increaseCreditLimit(BigDecimal amount) {
        if (hasGoodPaymentHistory()) {
            this.creditLimit = this.creditLimit.add(amount);
        } else {
            throw new IllegalStateException("Cannot increase credit limit - poor payment history");
        }
    }

    private boolean hasGoodPaymentHistory() {
        long paidOrders = orders.stream().filter(Order::isPaid).count();
        long totalOrders = orders.size();
        return totalOrders > 0 && (double) paidOrders / totalOrders >= 0.95;
    }

    // Getters
    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    public String getName() { return name; }
    public String getEmail() { return email; }
    public CustomerType getType() { return type; }
    public List<Order> getOrders() { return new ArrayList<>(orders); }
    public BigDecimal getCreditLimit() { return creditLimit; }
}

File: CSharp/Enterprise/DomainLogic/DomainModel/Customer.cs
using System;

namespace Enterprise.DomainLogic.DomainModel;

/// <summary>
/// Domain Model - Customer entity with business logic.
/// </summary>
public class Customer
{
    public int Id { get; private set; }
    public string Name { get; private set; }
    public string Email { get; private set; }
    public decimal CreditLimit { get; private set; }
    public bool IsVip { get; private set; }

    public Customer(int id, string name, string email, decimal creditLimit, bool isVip = false)
    {
        if (string.IsNullOrWhiteSpace(name))
            throw new ArgumentException("Name is required", nameof(name));

        if (string.IsNullOrWhiteSpace(email))
            throw new ArgumentException("Email is required", nameof(email));

        if (creditLimit < 0)
            throw new ArgumentException("Credit limit cannot be negative", nameof(creditLimit));

        Id = id;
        Name = name;
        Email = email;
        CreditLimit = creditLimit;
        IsVip = isVip;
    }

    /// <summary>
    /// Business logic to check if customer can place an order.
    /// </summary>
    public bool CanPlaceOrder(decimal orderAmount)
    {
        return orderAmount <= CreditLimit;
    }

    /// <summary>
    /// Increases customer's credit limit.
    /// </summary>
    public void IncreaseCreditLimit(decimal amount)
    {
        if (amount <= 0)
            throw new ArgumentException("Amount must be positive", nameof(amount));

        CreditLimit += amount;
    }

    /// <summary>
    /// Promotes customer to VIP status.
    /// </summary>
    public void PromoteToVip()
    {
        if (!IsVip)
        {
            IsVip = true;
            CreditLimit *= 1.5m; // Increase credit limit by 50%
        }
    }
}
